File: fc-chat/demo/part3/ch3/03-2-message-system-add-channel/src/main/java/net/prostars/messagesystem/service/UserConnectionService.java
package net.prostars.messagesystem.service;

import jakarta.persistence.EntityNotFoundException;
import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;
import net.prostars.messagesystem.constant.UserConnectionStatus;
import net.prostars.messagesystem.dto.domain.InviteCode;
import net.prostars.messagesystem.dto.domain.User;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.projection.UserIdUsernameInviterUserIdProjection;
import net.prostars.messagesystem.entity.UserConnectionEntity;
import net.prostars.messagesystem.repository.UserConnectionRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class UserConnectionService {

  private static final Logger log = LoggerFactory.getLogger(UserConnectionService.class);

  private final UserService userService;
  private final UserConnectionLimitService userConnectionLimitService;
  private final UserConnectionRepository userConnectionRepository;

  public UserConnectionService(
      UserService userService,
      UserConnectionLimitService userConnectionLimitService,
      UserConnectionRepository userConnectionRepository) {
    this.userService = userService;
    this.userConnectionLimitService = userConnectionLimitService;
    this.userConnectionRepository = userConnectionRepository;
  }

  public List<User> getUsersByStatus(UserId userId, UserConnectionStatus status) {
    List<UserIdUsernameInviterUserIdProjection> usersA =
        userConnectionRepository.findByPartnerAUserIdAndStatus(userId.id(), status);
    List<UserIdUsernameInviterUserIdProjection> usersB =
        userConnectionRepository.findByPartnerBUserIdAndStatus(userId.id(), status);
    if (status == UserConnectionStatus.ACCEPTED) {
      return Stream.concat(usersA.stream(), usersB.stream())
          .map(item -> new User(new UserId(item.getUserId()), item.getUsername()))
          .toList();
    } else {
      return Stream.concat(usersA.stream(), usersB.stream())
          .filter(item -> !item.getInviterUserId().equals(userId.id()))
          .map(item -> new User(new UserId(item.getUserId()), item.getUsername()))
          .toList();
    }
  }

  public UserConnectionStatus getStatus(UserId inviterUserId, UserId partnerUserId) {
    return userConnectionRepository
            .findUserConnectionStatusByPartnerAUserIdAndPartnerBUserId(
                    Long.min(inviterUserId.id(), partnerUserId.id()),
                    Long.max(inviterUserId.id(), partnerUserId.id()))
            .map(status -> UserConnectionStatus.valueOf(status.getStatus()))
            .orElse(UserConnectionStatus.NONE);
  }

  public Pair<Optional<UserId>, String> invite(UserId inviterUserId, InviteCode inviteCode) {
    Optional<User> partner = userService.getUser(inviteCode);
    if (partner.isEmpty()) {
      log.info("Invalid invite code. {}, from {}", inviteCode, inviterUserId);
      return Pair.of(Optional.empty(), "Invalid invite code.");
    }

    UserId partnerUserId = partner.get().userId();
    String partnerUsername = partner.get().username();
    if (partnerUserId.equals(inviterUserId)) {
      return Pair.of(Optional.empty(), "Can't self invite.");
    }

    UserConnectionStatus userConnectionStatus = getStatus(inviterUserId, partnerUserId);
    return switch (userConnectionStatus) {
      case NONE, DISCONNECTED -> {
        if (userService
            .getConnectionCount(inviterUserId)
            .filter(count -> count >= userConnectionLimitService.getLimitConnections())
            .isPresent()) {
          yield Pair.of(Optional.empty(), "Connection limit reached.");
        }

        Optional<String> inviterUsername = userService.getUsername(inviterUserId);
        if (inviterUsername.isEmpty()) {
          log.warn("InviteRequest failed.");
          yield Pair.of(Optional.empty(), "InviteRequest failed.");
        }
        try {
          setStatus(inviterUserId, partnerUserId, UserConnectionStatus.PENDING);
          yield Pair.of(Optional.of(partnerUserId), inviterUsername.get());
        } catch (Exception ex) {
          log.error("Set pending failed. cause: {}", ex.getMessage());
          yield Pair.of(Optional.empty(), "InviteRequest failed.");
        }
      }
      case ACCEPTED ->
          Pair.of(Optional.of(partnerUserId), "Already connected with " + partnerUsername);
      case PENDING, REJECTED -> {
        log.info(
            "{} invites {} but does not deliver the invitation request.",
            inviterUserId,
            partnerUsername);
        yield Pair.of(Optional.of(partnerUserId), "Already invited to " + partnerUsername);
      }
    };
  }

  public Pair<Optional<UserId>, String> accept(UserId acceptorUserId, String inviterUsername) {
    Optional<UserId> userId = userService.getUserId(inviterUsername);
    if (userId.isEmpty()) {
      return Pair.of(Optional.empty(), "Invalid username.");
    }
    UserId inviterUserId = userId.get();

    if (acceptorUserId.equals(inviterUserId)) {
      return Pair.of(Optional.empty(), "Can't self accept.");
    }

    if (getInviterUserId(acceptorUserId, inviterUserId)
        .filter(invitationSenderUserId -> invitationSenderUserId.equals(inviterUserId))
        .isEmpty()) {
      return Pair.of(Optional.empty(), "Invalid username.");
    }

    UserConnectionStatus userConnectionStatus = getStatus(inviterUserId, acceptorUserId);
    if (userConnectionStatus == UserConnectionStatus.ACCEPTED) {
      return Pair.of(Optional.empty(), "Already connected.");
    }
    if (userConnectionStatus != UserConnectionStatus.PENDING) {
      return Pair.of(Optional.empty(), "Accept failed.");
    }

    Optional<String> acceptorUsername = userService.getUsername(acceptorUserId);
    if (acceptorUsername.isEmpty()) {
      log.error("Invalid userId. userId: {}", acceptorUserId);
      return Pair.of(Optional.empty(), "Accept failed.");
    }

    try {
      userConnectionLimitService.accept(acceptorUserId, inviterUserId);
      return Pair.of(Optional.of(inviterUserId), acceptorUsername.get());
    } catch (IllegalStateException ex) {
      return Pair.of(Optional.empty(), ex.getMessage());
    } catch (Exception ex) {
      log.error("Accept failed. cause: {}", ex.getMessage());
      return Pair.of(Optional.empty(), "Accept failed.");
    }
  }

  public Pair<Boolean, String> reject(UserId senderUserId, String inviterUsername) {
    return userService
        .getUserId(inviterUsername)
        .filter(inviterUserId -> !inviterUserId.equals(senderUserId))
        .filter(
            inviterUserId ->
                getInviterUserId(inviterUserId, senderUserId)
                    .filter(invitationSenderUserId -> invitationSenderUserId.equals(inviterUserId))
                    .isPresent())
        .filter(
            inviterUserId -> getStatus(inviterUserId, senderUserId) == UserConnectionStatus.PENDING)
        .map(
            inviterUserId -> {
              try {
                setStatus(inviterUserId, senderUserId, UserConnectionStatus.REJECTED);
                return Pair.of(true, inviterUsername);
              } catch (Exception ex) {
                log.error("Set rejected failed. cause: {}", ex.getMessage());
                return Pair.of(false, "Reject failed.");
              }
            })
        .orElse(Pair.of(false, "Reject failed."));
  }

  public Pair<Boolean, String> disconnect(UserId senderUserId, String partnerUsername) {
    return userService
        .getUserId(partnerUsername)
        .filter(partnerUserId -> !senderUserId.equals(partnerUserId))
        .map(
            partnerUserId -> {
              try {
                UserConnectionStatus userConnectionStatus = getStatus(senderUserId, partnerUserId);
                if (userConnectionStatus == UserConnectionStatus.ACCEPTED) {
                  userConnectionLimitService.disconnect(senderUserId, partnerUserId);
                  return Pair.of(true, partnerUsername);
                } else if (userConnectionStatus == UserConnectionStatus.REJECTED
                    && getInviterUserId(senderUserId, partnerUserId)
                        .filter(inviterUserId -> inviterUserId.equals(partnerUserId))
                        .isPresent()) {
                  setStatus(senderUserId, partnerUserId, UserConnectionStatus.DISCONNECTED);
                  return Pair.of(true, partnerUsername);
                }
              } catch (Exception ex) {
                log.error("Disconnect failed. cause: {}", ex.getMessage());
              }
              return Pair.of(false, "Disconnect failed.");
            })
        .orElse(Pair.of(false, "Disconnect failed."));
  }

  private Optional<UserId> getInviterUserId(UserId partnerAUserId, UserId partnerBUserId) {
    return userConnectionRepository
        .findInviterUserIdByPartnerAUserIdAndPartnerBUserId(
            Long.min(partnerAUserId.id(), partnerBUserId.id()),
            Long.max(partnerAUserId.id(), partnerBUserId.id()))
        .map(inviterUserId -> new UserId(inviterUserId.getInviterUserId()));
  }

  @Transactional
  private void setStatus(
      UserId inviterUserId, UserId partnerUserId, UserConnectionStatus userConnectionStatus) {
    if (userConnectionStatus == UserConnectionStatus.ACCEPTED) {
      throw new IllegalArgumentException("Can't set to accepted.");
    }

    userConnectionRepository.save(
        new UserConnectionEntity(
            Long.min(inviterUserId.id(), partnerUserId.id()),
            Long.max(inviterUserId.id(), partnerUserId.id()),
            userConnectionStatus,
            inviterUserId.id()));
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/service/UserConnectionService.kt
package com.yscorp.withpush.messagesystem.service

import com.yscorp.withpush.messagesystem.constant.UserConnectionStatus
import com.yscorp.withpush.messagesystem.dto.domain.InviteCode
import com.yscorp.withpush.messagesystem.dto.domain.User
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.entity.UserConnectionEntity
import com.yscorp.withpush.messagesystem.repository.UserConnectionRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class UserConnectionService(
    private val userService: UserService,
    private val userConnectionLimitService: UserConnectionLimitService,
    private val userConnectionRepository: UserConnectionRepository
) {

    @Transactional(readOnly = true)
    fun getUsersByStatus(userId: UserId, status: UserConnectionStatus): List<User> {
        val usersA = userConnectionRepository.findByPartnerAUserIdAndStatus(userId.id, status)
        val usersB = userConnectionRepository.findByPartnerBUserIdAndStatus(userId.id, status)

        return (usersA + usersB)
            .filter { status != UserConnectionStatus.ACCEPTED || it.inviterUserId != userId.id }
            .map { User(UserId(it.userId), it.username) }
    }

    fun getStatus(inviterUserId: UserId, partnerUserId: UserId): UserConnectionStatus {
        val a = inviterUserId.id
        val b = partnerUserId.id

        return userConnectionRepository
            .findUserConnectionStatusByPartnerAUserIdAndPartnerBUserId(minOf(a, b), maxOf(a, b))
            ?.status
            ?.let(UserConnectionStatus::valueOf)
            ?: UserConnectionStatus.NONE
    }

    @Transactional(readOnly = true)
    fun countConnectionStatus(senderUserId: UserId, partnerUserIds: List<UserId>, status: UserConnectionStatus?): Long {
        val ids = partnerUserIds.map { it.id }
        return userConnectionRepository.countByPartnerAUserIdAndPartnerBUserIdInAndStatus(senderUserId.id, ids, status) +
            userConnectionRepository.countByPartnerBUserIdAndPartnerAUserIdInAndStatus(senderUserId.id, ids, status)
    }

    fun invite(inviterUserId: UserId, inviteCode: InviteCode): Pair<UserId?, String> {
        val partner = userService.getUser(inviteCode) ?: return null to "Invalid invite code."

        val partnerUserId = partner.userId
        val partnerUsername = partner.username

        if (partnerUserId == inviterUserId) return null to "Can't self invite."

        return when (val status = getStatus(inviterUserId, partnerUserId)) {
            UserConnectionStatus.NONE, UserConnectionStatus.DISCONNECTED -> {
                val isLimitReached = userService.getConnectionCount(inviterUserId)
                    ?.let { it >= userConnectionLimitService.limitConnections } ?: false

                if (isLimitReached) return null to "Connection limit reached."

                val inviterUsername = userService.getUsername(inviterUserId) ?: return null to "InviteRequest failed."

                try {
                    setStatus(inviterUserId, partnerUserId, UserConnectionStatus.PENDING)
                    partnerUserId to inviterUsername
                } catch (ex: Exception) {
                    log.error("Set pending failed. cause: {}", ex.message)
                    null to "InviteRequest failed."
                }
            }

            UserConnectionStatus.ACCEPTED -> partnerUserId to "Already connected with $partnerUsername"

            UserConnectionStatus.PENDING, UserConnectionStatus.REJECTED -> {
                log.info("{} invites {} but does not deliver the invitation request.", inviterUserId, partnerUsername)
                partnerUserId to "Already invited to $partnerUsername"
            }
        }
    }

    @Transactional
    fun accept(acceptorUserId: UserId, inviterUsername: String): Pair<UserId?, String> {
        val inviterUserId = userService.getUserId(inviterUsername) ?: return null to "Invalid username."

        if (acceptorUserId == inviterUserId) return null to "Can't self accept."

        val actualInviter = getInviterUserId(acceptorUserId, inviterUserId)
        if (actualInviter != inviterUserId) return null to "Invalid username."

        val status = getStatus(inviterUserId, acceptorUserId)
        if (status == UserConnectionStatus.ACCEPTED) return null to "Already connected."
        if (status != UserConnectionStatus.PENDING) return null to "Accept failed."

        val acceptorUsername = userService.getUsername(acceptorUserId) ?: return null to "Accept failed."

        return try {
            userConnectionLimitService.accept(acceptorUserId, inviterUserId)
            inviterUserId to acceptorUsername
        } catch (ex: IllegalStateException) {
            null to ex.message!!
        } catch (ex: Exception) {
            log.error("Accept failed. cause: {}", ex.message)
            null to "Accept failed."
        }
    }

    fun reject(senderUserId: UserId, inviterUsername: String): Pair<Boolean, String> {
        val inviterUserId = userService.getUserId(inviterUsername) ?: return false to "Reject failed."

        if (inviterUserId == senderUserId) return false to "Reject failed."

        val actualInviter = getInviterUserId(inviterUserId, senderUserId)
        if (actualInviter != inviterUserId) return false to "Reject failed."

        if (getStatus(inviterUserId, senderUserId) != UserConnectionStatus.PENDING) {
            return false to "Reject failed."
        }

        return try {
            setStatus(inviterUserId, senderUserId, UserConnectionStatus.REJECTED)
            true to inviterUsername
        } catch (ex: Exception) {
            log.error("Set rejected failed. cause: {}", ex.message)
            false to "Reject failed."
        }
    }

    @Transactional
    fun disconnect(senderUserId: UserId, partnerUsername: String): Pair<Boolean, String> {
        val partnerUserId = userService.getUserId(partnerUsername) ?: return false to "Disconnect failed."
        if (senderUserId == partnerUserId) return false to "Disconnect failed."

        return try {
            val status = getStatus(senderUserId, partnerUserId)
            if (status == UserConnectionStatus.ACCEPTED) {
                userConnectionLimitService.disconnect(senderUserId, partnerUserId)
                true to partnerUsername
            } else if (
                status == UserConnectionStatus.REJECTED &&
                getInviterUserId(senderUserId, partnerUserId) == partnerUserId
            ) {
                setStatus(senderUserId, partnerUserId, UserConnectionStatus.DISCONNECTED)
                true to partnerUsername
            } else {
                false to "Disconnect failed."
            }
        } catch (ex: Exception) {
            log.error("Disconnect failed. cause: {}", ex.message)
            false to "Disconnect failed."
        }
    }

    private fun getInviterUserId(partnerAUserId: UserId, partnerBUserId: UserId): UserId? {
        return userConnectionRepository
            .findInviterUserIdByPartnerAUserIdAndPartnerBUserId(
                minOf(partnerAUserId.id, partnerBUserId.id),
                maxOf(partnerAUserId.id, partnerBUserId.id)
            )
            ?.inviterUserId
            ?.let(::UserId)
    }

    @Transactional
    fun setStatus(inviterUserId: UserId, partnerUserId: UserId, status: UserConnectionStatus) {
        require(status != UserConnectionStatus.ACCEPTED) { "Can't set to accepted." }

        userConnectionRepository.save(
            UserConnectionEntity(
                minOf(inviterUserId.id, partnerUserId.id),
                maxOf(inviterUserId.id, partnerUserId.id),
                status,
                inviterUserId.id
            )
        )
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(UserConnectionService::class.java)
    }
}
